feat(pipeline): multi-branch pipelines

jstrunk opened the pull request:
If your team follows [GitHub flow](https://docs.github.com/en/get-started/quickstart/github-flow), the pipelines module can optionally help you create independent environments to test and validate changes before merging. When you create a new branch, it will automatically create a new pipeline stack and any stages you configure. Once you have finished testing and delete the branch, it will automatically clean up the stacks created in the branch's environment.

This feature is based on the architecture in [Multi-branch pipeline management and infrastructure deployment using AWS CDK Pipelines](https://aws.amazon.com/blogs/devops/multi-branch-pipeline-management-and-infrastructure-deployment-using-aws-cdk-pipelines/) . EventBridge Rules trigger Lambda functions on CodeCommit repo branch creation and deletion. The Lambda function for creation sets the `BRANCH` environment variable and runs a CodeBuild project that creates the pipeline stack for the branch. It is up to the developer to give their pipeline stack and stages unique names per branch. The examples show how to use the `BRANCH` variable to do so. The multi-branch stacks are tagged with the branch name and repository name. The Lambda for branch deletion lists all stacks matching those tags and deletes them.

I tested this change using a Typescript pipeline project.

Resolves #302

jstrunk committed:
feat(pipeline): multi-branch pipelines

If your team follows [GitHub flow](https://docs.github.com/en/get-started/quickstart/github-flow),
the pipelines module can optionally help you create independent environments to test and validate
changes before merging. When you create a new branch, it will automatically create a new pipeline
stack and any stages you configure. Once you have finished testing and delete the branch, it will
automatically clean up the stacks created in the branch's environment.

EventBridge Rules trigger Lambda functions on CodeCommit repo branch creation and deletion. The
Lambda function for creation sets the `BRANCH` environment variable and runs a CodeBuild project
that creates the pipeline stack for the branch. It is up to the developer to give their pipeline
stack and stages unique names per branch. The examples show how to use the `BRANCH` variable to do
so. The multi-branch stacks are tagged with the branch name and repository name. The Lambda for
branch deletion lists all stacks matching those tags and deletes them.

I tested this change using a Typescript pipeline project.

cogwirrel commented on packages/pipeline/README.md:
```diff
@@ -46,4 +46,82 @@ npx projen new --from aws-prototyping-sdk pdk-pipeline-java
+
+### Pipeline Definition
+
+When you define you pipeline, you define which stages get created for a given branch and how to name your stacks uniquely. `PipelineStack` must be included.
```
Nit: *your pipeline :)

cogwirrel commented on packages/pipeline/test/pdk-pipeline.test.ts:
```diff
@@ -195,4 +195,191 @@ describe("PDK Pipeline Unit Tests", () => {
+    app.synth();
+    expect(app.nagResults().length).toEqual(0);
+    expect(Template.fromStack(stack)).toMatchSnapshot();
+    delete process.env.BRANCH;
```
It might be worth cleaning up the environment in a `finally` block to reduce the chances of a failure in one test affecting another :)

jstrunk replied:
I implemented an `afterEach` function to restore `process.env`.

cogwirrel commented on packages/pipeline/README.md:
```diff
@@ -46,4 +46,82 @@ npx projen new --from aws-prototyping-sdk pdk-pipeline-java
+this.pipeline = new PDKPipeline(this, "ApplicationPipeline", {
+  primarySynthDirectory: "packages/backend/cdk.out",
+  repositoryName: this.node.tryGetContext("repositoryName") || "monorepo",
+  branchNamePrefixes: [""],
```
It might be a little more readable if we exposed a constant like `PDKPipeline.ALL_BRANCHES`? `[""]` makes logical sense but is a little harder to reason about at first glance :)

cogwirrel commented on packages/pipeline/src/utils.ts:
```diff
@@ -0,0 +1,17 @@
+
+import { DEFAULT_BRANCH_NAME } from "./pdk-pipeline";
+
+export class PipelineUtils {
```
Nit: it might be neater to have these as static members of the `PDKPipeline` construct rather than separate them out as they're quite closely related :)

cogwirrel commented on packages/pipeline/src/feature-branches.ts:
```diff
@@ -0,0 +1,361 @@
+        },
+        buildSpec: BuildSpec.fromObjectToYaml({
+          version: "0.2",
+          phases: {
+            install: {
+              commands: [
+                "npm install -g aws-cdk",
+                "yarn install --frozen-lockfile || npx projen && yarn install --frozen-lockfile",
+              ],
+            },
+            build: {
+              commands: [
+                "npx nx run-many --target=build --all",
+                `cd ${props.cdkSrcDir}`,
+                "npx cdk synth",
+                "npx cdk deploy --require-approval=never",
+              ],
+            },
+          },
```
The user can pass custom install/build commands to `PDKPipeline`, and so I think we want to honour those here too?

jstrunk replied:
I think it will work to replace only `"npx nx run-many --target=build --all"` with `commands` from `PDKPipeline`. The user wouldn't pass installation commands to `commands` in `synthShellStepPartialProps`.

cogwirrel replied:
You can pass `installCommands` to `synthShellStepPartialProps`, which is particularly useful if you're using a different package manager to yarn, like pnpm :) So I think we definitely want to honour what's configured in `PDKPipeline` where possible.

cogwirrel replied:
You can pass `installCommands` to `synthShellStepPartialProps`, which is particularly useful if you're using a different package manager to yarn, like pnpm :) So I think we definitely want to honour what's configured in `PDKPipeline` where possible.

cogwirrel commented on packages/pipeline/src/feature-branches.ts:
```diff
@@ -0,0 +1,361 @@
+   *   repositoryName: 'my-repo',
+   *   branchNamePrefixes: [], // or simply exclude this line
+   * }
+   * @default undefined
+   */
+  readonly branchNamePrefixes: string[];
```
Nit: looks like this prop is required so doesn't have a default here :)

cogwirrel commented on packages/pipeline/samples/typescript/src/pipeline.ts:
```diff
@@ -1,19 +1,27 @@
 const app = PDKNag.app();
 
-const pipelineStack = new PipelineStack(app, 'PipelineStack', {
+const branchPrefix =
+  process.env.BRANCH &&
+  process.env.BRANCH !==
+    (app.node.tryGetContext('defaultBranchName') || PipelineUtils.defaultBranchName)
+    ? PipelineUtils.normalizeBranchName(process.env.BRANCH) + '-'
+    : '';
```
It might be nicer to abstract away the environment variable from users, and provide some helpers like `PDKPipeline.getBranchPrefix(app)` and `PDKPipeline.isDefaultBranch(app)`?

cogwirrel commented on packages/pipeline/src/feature-branches.ts:
```diff
@@ -0,0 +1,361 @@
+      eventPattern: {
+        detail: {
+          referenceType: ["branch"],
+          referenceName: props.branchNamePrefixes.map((prefix) => {
+            return { prefix };
+          }),
```
Nit: `.map((prefix) => ({ prefix }))`

cogwirrel reviewed (commented):
Very cool feature, nice job! 🚀

JeremyJonas commented:
@jstrunk What is the status of this PR?

jstrunk commented:
I am going to implement the comments, but I didn't have time last week. I
expect to make time for it this week.

On Sun, Apr 30, 2023, 22:36 Jeremy Jonas ***@***.***> wrote:

> @jstrunk <https://github.com/jstrunk> What is the status of this PR?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/aws/aws-prototyping-sdk/pull/364#issuecomment-1529273037>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAAUYCUKSEX5XNAH4J5UF3LXD4ORFANCNFSM6AAAAAAXHL2OEA>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

jstrunk commented:
I've addressed the comments. Tomorrow I will update the samples.

JeremyJonas committed:
fix(nx-monorepo): remove projen script

This script is not used and causes recursive execution
for other scripts like "yarn projen build".

agdimech committed:
Merge branch 'mainline' into feature/pipeline-feature-branches

JeremyJonas committed:
fix(nx-monorepo): use exec for yarn2 and refactor npm utils

agdimech committed:
fix: bump dependencies to use TS 5.X (#396)

cogwirrel commented on packages/pipeline/src/feature-branches.ts:
```diff
@@ -0,0 +1,368 @@
+            install: {
+              commands: [
+                "npm install -g aws-cdk",
+                "yarn install --frozen-lockfile || npx projen && yarn install --frozen-lockfile",
```
We should allow the user to override the install commands too as not everyone will use `yarn`, ie some may use `pnpm` or `npm` :) They're currently overideable in `PDKPipeline` via `synthShellStepPartialProps` so I think we can pass whatever value that has if present, eg `synthShellStepPartialProps?.installCommands ?? ['npm install -g aws-cdk', 'yarn ....']`

jstrunk replied:
I modified the construct to take `synthShellStepPartialProps` directly and get build command and install commands there.

cogwirrel commented on packages/pipeline/src/feature-branches.ts:
```diff
@@ -0,0 +1,368 @@
+                ...buildCommands,
+                `cd ${props.cdkSrcDir}`,
+                "npx cdk synth",
+                "npx cdk deploy --require-approval=never",
```
Technically depending on the package manager one might not always use `npx` (eg if you're using pnpm the command might be `pnpm dlx`). I wonder if we should also allow the user to override default deploy commands?

jstrunk replied:
I added a `cdkCommand` prop to customize this.

cogwirrel commented on packages/pipeline/src/pdk-pipeline.ts:
```diff
@@ -87,9 +136,66 @@ export interface PDKPipelineProps extends CodePipelineProps {
+   * }
+   * @returns True if the current branch is the default branch.
+   */
+  public static isDefaultBranch(
```
Nice! 😄

cogwirrel reviewed (commented):
Looks great, couple of minor comments :)

agdimech committed:
feat(type-safe-api): add a DefaultApiClientProvider which handles shared context (#401)

jstrunk committed:
feat: allow customization of CodeBuild project

jstrunk commented:
I've implemented the latest comments.

cogwirrel reviewed (approved):
Great stuff! Thanks!

agdimech commented:
@jstrunk - Hi Jeff. I am noticing a lot of conflicts and files that have changed that are not directly apart of this PR (due to the large amount of structural PRs that have been merged recently and how the rebase has occurred).

Are you able to just cherry pick your specific changes into a fresh branch? Otherwise I am more than happy to do this on your behalf if you prefer?

Apart from that all your code looks good and can be merged once we get the commit/merges cleaned up :)

jstrunk commented:
> @jstrunk - Hi Jeff. I am noticing a lot of conflicts and files that have changed that are not directly apart of this PR (due to the large amount of structural PRs that have been merged recently and how the rebase has occurred).
> 
> Are you able to just cherry pick your specific changes into a fresh branch? Otherwise I am more than happy to do this on your behalf if you prefer?
> 
> Apart from that all your code looks good and can be merged once we get the commit/merges cleaned up :)

That makes sense. I don't think that should be to much trouble. The files that belong in the PR can be copied over whole.

If I do that, how should I associate that new branch with this pr?

agdimech commented:
> > @jstrunk - Hi Jeff. I am noticing a lot of conflicts and files that have changed that are not directly apart of this PR (due to the large amount of structural PRs that have been merged recently and how the rebase has occurred).
> > Are you able to just cherry pick your specific changes into a fresh branch? Otherwise I am more than happy to do this on your behalf if you prefer?
> > Apart from that all your code looks good and can be merged once we get the commit/merges cleaned up :)
> 
> That makes sense. I don't think that should be to much trouble. The files that belong in the PR can be copied over whole.
> 
> If I do that, how should I associate that new branch with this pr?

You can just open a new PR, abandon this one and just add a reference to this PR in the description of the new PR for traceability :)

jstrunk commented:
Please continue this discussion in #403 .
